contrib/cf-diag.sh

```diff
@@ -0,0 +1,240 @@
+# (It will also fail with coreutils 8.6, but 8.5 and 8.7+ are fine:
+# see https://unix.stackexchange.com/q/3348/135943.)  :)
+
+desc="cf-agent ran in last 15 minutes"
+if [ "$(($(date +%s)-$(stat -c %Y /var/cfengine/promise_summary.log)))" -lt 900 ]; then
+  succeed "$desc"
+else
+  fail "$desc"
+fi
+
+# HUB CHECKS
+
+desc="Bootstrapped to hub (policy_server.dat exists)"
+if [ -f /var/cfengine/policy_server.dat ]; then
+  succeed "$desc"
+  hub_ip="$(cat /var/cfengine/policy_server.dat)"
+  printf '# Hub IP is %s\n' "$hub_ip"
+else
+  fail "$desc"
+  bail 'Not connected to any hub.'
+fi
+
+desc="Hub is connecting to host"
+if cf-key -s | grep ^Incoming | grep -qF "$hub_ip"; then
+  succeed "$desc"
+
+  hub_lastseen="$(cf-key -s |grep -F "$hub_ip" |awk '/^Incoming/{print $4, $5, $6, $7, $8}')"
+  hub_lastseen_in_seconds="$(date +%s -d "$hub_lastseen")"
+  now="$(date +%s)"
+  seconds_since_conn="$((now-hub_lastseen_in_seconds))"
+
+  desc="Hub connected to host within last 15 minutes"
+  if [ "$seconds_since_conn" -lt 900 ]; then
+    succeed "$desc"
+  else
+    fail "$desc"
+  fi
+  printf '# Has been %d seconds since last connection\n' "$seconds_since_conn"
+
+else
+  # Hub connection fail
+  fail "$desc"
+fi
+
+### SYSTEM CHECKS
+
+disk_usage="$(df -P / | awk '/\//{print $5}')"
+
+desc="Filesystem has free space"
+if [ "$disk_usage" = '100%' ]; then
+  fail "$desc"
+  bail "Can't continue with full disk"
+else
+  succeed "$desc"
+  printf '# Disk usage is %s\n' "$disk_usage"
+fi
+
+desc="Filesystem is writable"
+if touch /tmp/cf_diag_test 2>/dev/null; then
+  succeed "$desc"
+  rm /tmp/cf_diag_test
+else
+  fail "$desc"
+  bail "Can't continue with read-only filesystem"
+fi
+
+### POLICY CHECKS
+
+for f in promises.cf update.cf; do
+  desc="$f exists"
+  if [ -f /var/cfengine/inputs/"$f" ]; then
+    succeed "$desc"
+  else
+    fail "$desc"
+    bail "No $f"
+  fi
+done
+
+# The tests below this point may take a long time to run,
+# depending on how efficiently your CFEngine policy is written.
+
+for f in promises.cf update.cf; do
+  desc="$f is runnable"
+  if cf-agent -f "$f" >/dev/null 2>&1; then
+    succeed "$desc"
+  else
+    fail "$desc"
+    bail "$f not runnable"
+  fi
+done
+
+desc="Update run fixes changes to policy"
+# Running update policy (after deleting flag file)
+# should repair alteration of promises.cf
+echo '#cf-diag_comment' >> /var/cfengine/inputs/promises.cf
+rm -f /var/cfengine/inputs/cf_promises_validated
+cf-agent -K -f update.cf >/dev/null 2>&1
+if grep -q cf-diag_comment /var/cfengine/inputs/promises.cf; then
+  fail "$desc"
+else
+  succeed "$desc"
+fi
+
+############################### CUSTOM TESTS ##################################
+
+# Add custom tests here (if you really need them).
+# Use the general form:
+#
+#   desc="Name of test (descriptive string)"
+#   if some-command-that-should-succeed; then
+#     succeed "$desc"
+#   else
+#     fail "$desc"
+#   fi
```

Comment:
Nice stuff! But what about replacing `if (condition); then succeed "$desc" else fail "$desc"; fi` with something like 
```
(condition)
check $? "$desc"
```
where
```
check() {
  if [ $1 = 0 ]; then
    succeed "$2"
  else
    fail "$2"
  fi
}
```
You can also add third optional argument to bail if the test fails, and/or add `return $1` at the end to make it chain-able like this:
```
check $? "$desc" || bail "Fatal error"
```

Reply:
I did consider that approach.  However, not all tests actually follow that pattern.  Some tests are conditional on the success or failure of a different test, as you can see.

If there's one thing I try to avoid like the plague in shell scripts, it's building up frameworks that don't actually match the requirements.  Shell scripts are simple enough that they don't *need* wrappers.

Reply:
Yes, like nested `if`s...

But maybe for simple cases?